packages/kbn-apm-synthtrace/src/lib/utils/create_metric_aggregator_factory.ts

```diff
@@ -84,6 +100,92 @@ export function createMetricAggregatorFactory<TFields extends Fields>() {
         }
 
         const timestamp = event['@timestamp']!;
+        function generateTransactionMetric(service: ServiceMapValue) {
```

Comment:
this is too specific imho, we should be able to deal with this by adding config options rather than having specific function calls and if/else branches.

Reply:
I think what it generally comes down to is that every time you check the metrics Map for an entry for key x, and it's missing, you also check the size of the metrics Map. If it is above a certain threshold, you write to the overflow bucket.

Reply:
> this is too specific imho, we should be able to deal with this by adding config options rather than having specific function calls and if/else branches.

TBH, when i studied the code, i was really awe struck with the implementation thats its so generic and neat. The factory literally has 0 knowledge about the event its processing and individual aggregators computing the individual metrics documents.

But for the given requirement, i had to pollute this generic logic as this factory is the only place which maintains the closure and its state which we need to compute the count.

> I think what it generally comes down to is that every time you check the metrics Map for an entry for key x, and it's missing, you also check the size of the metrics Map. If it is above a certain threshold, you write to the overflow bucket.

I actually started with this exact logic and after implementation and running the code, realised that the `metrics` Map is not the right place to check for `size` to calculate overflow for `transactions`

Reason being - The `metrics` map stores all transactions for all services together in one place for a given `flushInterval`. This means if we have, 10 services and each with 20 transactions, this will generate 200 metrics inside the `metrics` Map. Now for a given `transaction overflow count` set at 15, this would means it would generate a `_other` bucket and push remaining 185 transactions inside that bucket. We don't want this.

What we actually need is to generate `_other` bucket for `each service`.
So that for 20 services, each service will have a `_other` bucket with the bucket size of 5.
Hence i created the `serviceListMap` which tracks this count per service and once the size is reached, it created the `_other` bucket for that service.

```
const serviceListMap: Map<string, ServiceMapValue> = new Map();
```

Now the below code checks for every incoming event, if its a transaction metrics and the service exists in the map or not.
If it does not exists, it create a service object where we maintain the count of each transaction and overflowKey (which will only be assigned if the service name is `_other`) and then process the transaction

```
function writeTransactionMetric() {
    const existingService = serviceListMap.get(event['service.name']);
    if (existingService) {
        generateTransactionMetric(existingService);
     } else {
         const newService: ServiceMapValue = {
           transactionCount: 0,
           overflowKey: null,
         };
      serviceListMap.set(event['service.name'], newService);
      generateTransactionMetric(newService);
     }
}
```

Reply:
> TBH, when i studied the code, i was really awe struck with the implementation thats its so generic and neat. The factory literally has 0 knowledge about the event its processing and individual aggregators computing the individual metrics documents.

I'll frame this one, thank you 😛 

> Reason being - The metrics map stores all transactions for all services together in one place for a given flushInterval. This means if we have, 10 services and each with 20 transactions, this will generate 200 metrics document inside the metrics Map. Now for a given transaction overflow count = 15, this would means it would generate a _other bucket and push remaining 185 transactions inside that bucket. We don't want this.

I see. But I think we can still achieve that without adding aggregator-specific logic into the aggregator factory. I think it boils down to additional grouping keys, of which there might be multiple, and then per each grouping key, there would be a limit. E.g., you'd have `groups: [ { keys: [ 'service.name' ], limit: 100 }, { keys: [ 'service.name', 'transaction.name' ], limit: 100 ]`. I haven't really thought the implementation through but I think that if we can define generic config we can also have a generic implementation. WDYT?

Reply:
> I think it boils down to additional grouping keys, of which there might be multiple, and then per each grouping key, there would be a limit.

This sounds good. Let me play with it to see how the exact implementation would look like.

Reply:
@dgieselaar I tried playing with the approach. I don't see a way to achieve this with generic configs.
But what i found was, we can add one more method to all the aggregators call `group` similar to `init`, `reduce`, `filter`. This function will take care of setting metrics name and count values.

So the factory will do something like this which is generic and it currently does for every event

```
// all existing stuff
const state = {}

init()
reduce()
.......
// new stuff
group(state)
.......
```

This way every `group` function in the aggregators can have aggregator specific logic and the factory can be clean. The factory will still have to maintain the state variable for count

Reply:
@dgieselaar Finally i was able to keep the factory clean 😉